sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TestMethodShouldNotBeIgnored.cs

```diff
@@ -51,46 +53,56 @@ public sealed class TestMethodShouldNotBeIgnored : SonarDiagnosticAnalyzer
 
         protected override void Initialize(SonarAnalysisContext context)
         {
-            context.RegisterSyntaxNodeActionInNonGenerated(
-                c =>
-                {
-                    if (!c.IsTest())
-                    {
-                        return;
-                    }
-
-                    var methodSymbol = c.SemanticModel.GetDeclaredSymbol(c.Node);
-                    if (methodSymbol == null)
-                    {
-                        return;
-                    }
-
-                    var attributes = methodSymbol.GetAttributes();
-                    if (!attributes.Any(a => a.AttributeClass.IsAny(TrackedTestMethodAttributes)))
-                    {
-                        return;
-                    }
-
-                    var ignoreAttribute = attributes.FirstOrDefault(a =>
-                        a.AttributeClass.Is(KnownType.Microsoft_VisualStudio_TestTools_UnitTesting_IgnoreAttribute));
-                    if (ignoreAttribute == null)
-                    {
-                        return;
-                    }
-
-                    var ignoreAttributeSyntax = ignoreAttribute.ApplicationSyntaxReference.GetSyntax();
-
-                    var hasWorkItemAttribute = attributes.Any(a =>
-                        a.AttributeClass.Is(KnownType.Microsoft_VisualStudio_TestTools_UnitTesting_WorkItemAttribute));
-                    var hasTrailingComment = ignoreAttributeSyntax.Parent.GetTrailingTrivia()
-                        .Any(trivia => trivia.IsKind(SyntaxKind.SingleLineCommentTrivia));
-
-                    if (!hasWorkItemAttribute &&
-                        !hasTrailingComment)
-                    {
-                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, ignoreAttributeSyntax.GetLocation()));
-                    }
-                }, SyntaxKind.MethodDeclaration);
-        }
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+            {
+                var attribute = (AttributeSyntax)c.Node;
+                if (HasReasonPhrase(attribute) ||
+                    HasTrailingComment(attribute) ||
+                    !IsMsTestIgnoreAttribute(attribute, c.SemanticModel))
+                {
+                    return;
+                }
+
+                var attributeTarget = attribute.Parent?.Parent;
+                if (attributeTarget == null)
+                {
+                    return;
+                }
+
+                var testMethodOrClass = c.SemanticModel.GetDeclaredSymbol(attributeTarget);
+
+                if (testMethodOrClass != null &&
+                    testMethodOrClass.GetAttributes().Any(IsTestOrTestClassAttribute) &&

```

Comment:
Isn't it better to avoid querying `GetAttributes()` twice?

Reply:
@Evangelink I don't know, it returns ImmutableArray<T> which is value type and I cannot say if it is the same instance; I suppose it is returning a new instance because `GetAttributes()` is a method...